### PR TITLE
Adding board and grade models to schema.

### LINKED ISSUE #21: Adding Board and Grades to schema
Right now, the whole platform works assuming that it's for a single grade and a universal board. We need this additional constraints on Questions model

### PR DESCRIPTION
**Pull request by ishanisri**
This fixes #21 .
The hierarchy followed for the schema is Board->Grade->Subject.

### CODE REVIEW

**Comment by ishanisri** [comment]
According to this new schema, fixes would be required in the following places.

- uploading questions.

- creating a test paper.

- creating a customized paper.

- creating a generic paper.

- in subject splits.

**Comment by ishanisri** [comment]
The new format of the excel sheet to be followed while uploading questions can be found in the link below: [https://docs.google.com/spreadsheets/d/1cAmPmDPrHabwMWgT1ED7NPesVk3HqYyLuzkVG6Dj3fQ/edit?usp=sharing](Dummy Question Bank for Uploading Questions).

Mainly, two more columns have been added for board and grade before the subject name in this new format of question bank.

File: searchapp/models.py [hunk 1]
```diff
@@ -21,9 +21,37 @@ class Mentor(models.Model):
     def __str__(self):
         return self.username
 
+class Board(models.Model):
+    BOARD_TYPE_CHOICES = (
+        ('CBSE',  'CBSE'),
+        ('ICSE', 'ICSE'),
+    )
+    board = models.CharField(max_length=4,choices=BOARD_TYPE_CHOICES, default='CBSE',unique=True)
```
Comment (PaiAkshay998) [inline searchapp/models.py]: Stick to pep8. Spaces after comma.

File: searchapp/models.py [hunk 2]
```diff
@@ -21,9 +21,37 @@ class Mentor(models.Model):
...
+        self.min_value, self.max_value = min_value, max_value
+        super(MinMaxFloat, self).__init__(*args, **kwargs)
+
+    def formfield(self, **kwargs):
+        defaults = {'min_value': self.min_value, 'max_value' : self.max_value}
+        defaults.update(kwargs)
+        return super(MinMaxFloat, self).formfield(**defaults)
+
+class Grade(models.Model):
+    grade = MinMaxFloat(min_value=1, max_value=12, default=10)
```
Comment (PaiAkshay998) [inline searchapp/models.py]: Is this really needed? Can't we have server validation for this instead.

File: searchapp/utils/utils.py [hunk 3]
```diff
@@ -153,46 +153,52 @@ def convert_question_bank(question_bank_path):
...
+                chapter_no = int(row[3].value)
+                chapter_name = row[4].value
+                question_type = row[5].value
                 question_type = str(int(question_type)) if (type(question_type) == type(1.0) or type(question_type) == type(1)) else question_type.lower()
-                question_text = row[4].value
-                question_source = row[5].value
-                subject_to_chapter_to_question[subject][(chapter_no, chapter_name)][question_type].append((question_text, question_source))
+                question_text = row[6].value
+                question_source = row[7].value
+                board_to_grade_to_subject_to_chapter_to_question[board][grade][subject][(chapter_no, chapter_name)][question_type].append((question_text, question_source))
```
Comment (PaiAkshay998) [inline searchapp/utils/utils.py]: Name this `question_bank_dict` or something

File: searchapp/views.py [hunk 4]
```diff
@@ -140,9 +142,9 @@ def add_questions(request):
         if request.FILES:
             file_obj = request.FILES['datafile']
             default_dict = convert_question_bank(file_obj)
-            subject_to_chapter_to_question=default_to_regular(default_dict)
+            board_to_grade_to_subject_to_chapter_to_question=default_to_regular(default_dict)
```
Comment (PaiAkshay998) [inline searchapp/views.py]: Same as above

File: searchapp/models.py (hunk 2 above)
Comment (ishanisri) [reply searchapp/models.py]: This just helps in keeping the grades within this range even if someone tries to add new grades to the database from django admin panel.

**Comment by ishanisri** [comment]
For subject splits, the following functionalities have been updated :

- get chapters

- add chapters

- delete chapters

- display subject split

- add subject split

- delete subject split

I have shifted query searches in all these from subject_names to subject_ids to identify subjects of a particular board and grade, and the same for subject_splits. We should check for the same in other query searches also.

File: searchapp/models.py [hunk 5]
```diff
@@ -2,28 +2,63 @@
...
-    created_time = models.DateTimeField(auto_now_add=True, blank=True, null=True)
+    created_time = models.DateTimeField(
+        auto_now_add=True, blank=True, null=True)
 
     def __str__(self):
         return self.username
 
 
+class Board(models.Model):
+    BOARD_TYPE_CHOICES = (
```
Comment (PaiAkshay998) [inline searchapp/models.py]: Do not do this. Incase a new board needs to be added, this will force code changes. Ideally, you'd just want to be able to insert a new Board to the database and not require any code changes.

File: searchapp/migrations/0013_auto_20190812_1532.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# Generated by Django 2.1 on 2019-08-12 10:02
```
Comment (PaiAkshay998) [inline searchapp/migrations/0013_auto_20190812_1532.py]: I'm not sure if we should be pushing migrations. I'm not entirely sure about this.

File: searchapp/urls.py [hunk 7]
```diff
@@ -1,30 +1,40 @@
...
-   path('add_subject_split', views.add_subject_split, name='add_subject_split'),
-   path('get_test_paper', views.get_test_paper, name='get_test_paper'),
-   path('get_generic_paper', views.get_generic_paper, name='get_generic_paper'),
-   path('get_test_format', views.get_test_format, name='get_test_format'),
-   path('get_customize_paper',views.get_customize_paper, name='get_customize_paper'),
-   path('generate_optional_inputs',views.generate_optional_inputs,name = 'generate_optional_inputs'),
-   path('',views.home, name='home'),
-   path('contact',views.contact),
-   path('get_dummy_tracker', views.get_dummy_tracker),
+app_name = 'searchapp'
```
Comment (PaiAkshay998) [inline searchapp/urls.py]: Why are so many changes made to this file?

Comment (ishanisri) [reply searchapp/urls.py]: Oh, there are no changes as such. I just applied pep8 to this file so it is probably changes in spaces.

File: searchapp/models.py (hunk 5 above)
Comment (ishanisri) [reply searchapp/models.py]: This was to keep the naming and format of boards consistent in terms of capital and small letters and their abbreviations. I can change it if we do not need these checks.

File: searchapp/static/student_view.js [hunk 8]
```diff
@@ -162,6 +162,51 @@ function upload_click(e) {
...
+	$.ajax({
+		url: BASE_DIR + "/get_grades",
+		method : "get",
+		data: boardData,
+		headers: { "X-CSRFToken": csrftoken, crossOrigin: false},
+		success: function(response) {
+			$(`#${worksheetType}-grade-options-parent`).empty();
+			$(`#${worksheetType}-grade`).removeClass('hide-display').addClass('show-display');
+			let grades = response['grade_list'];
+			for (let i=0; i<grades.length; i++) {
```
Comment (PaiAkshay998) [inline searchapp/static/student_view.js]: Lets start moving away from conventional for loops. Try using `forEach` with arrow functions. Should improve readability;

File: searchapp/static/student_view.js [hunk 9]
```diff
@@ -162,6 +162,51 @@ function upload_click(e) {
...
+	$.ajax({
+		url: BASE_DIR + "/get_subjects",
+		method : "get",
+		data: gradeData,
+		headers: { "X-CSRFToken": csrftoken, crossOrigin: false},
+		success: function(response) {
+			$(`#${worksheetType}-subject-options-parent`).empty();
+			$(`#${worksheetType}-subject`).removeClass('hide-display').addClass('show-display');
+			let subjects = response['subject_list'];
+			for (let i=0; i<subjects.length; i++) {
```
Comment (PaiAkshay998) [inline searchapp/static/student_view.js]: Same here I guess.

File: searchapp/static/student_view.js [hunk 10]
```diff
@@ -190,26 +235,26 @@ function upload_click(e) {
...
-					
+
 					let subject_breakup = response['subject_breakup'];
 					$(`#subject_splits`).removeClass('show-display').addClass('hide-display');
 					$('h3').removeClass('show-display').addClass('hide-display');
 					$(`#chapter-operations`).removeClass('hide-display').addClass('show-display');
 					$('#add-subject-split').removeClass('hide-display').addClass('show-display');
-					
+
 					for (let i=0; i<subject_breakup.length; i++) {
```
Comment (PaiAkshay998) [inline searchapp/static/student_view.js]: Same

**Comment by sandy9999** [comment]
In the /chapter_and_split_view, when it comes to displaying subject splits, when I select cbse, it should show me all splits with that name. A split with cbse is not a distinct entity. I wouldn't want to clog my drop down with 10 instances of cbse, and select all 10.

What we need is to group splits with same name as one, and allow that name alone to appear in the display split dropdown as an option. And if that's selected, all instances should be displayed.

Screenshot from 2019-09-08 15-18-45
Screenshot from 2019-09-08 15-19-11

File: searchapp/utils/utils.py [hunk 11]
```diff
@@ -75,20 +77,23 @@ def convert_marker_data(marker_path, subject_breakup):
...
-                    questions_of_given_type.append((marks_secured, question_number, chapter_number))
+                    questions_of_given_type.append(
+                        (marks_secured, question_number, chapter_number))
                 questions_of_given_type.sort(key=lambda x: x[0], reverse=True)
                 questions_of_given_type = questions_of_given_type[:questions_to_attempt]
                 student_to_answer[student][question_type] = questions_of_given_type
                 question_type_start_row += total_questions
                 question_type = header_row[question_type_start_row].value
-    return student_to_answer, list(set([int(val.value) for val in chapter_row if type(val.value) == type(1.0)]))
+    return student_to_answer, list(set([int(val.value) for val in chapter_row if type(val.value) == type(1)]))
```
Comment (PaiAkshay998) [inline searchapp/utils/utils.py]: `type(1)` and `type(1.0)` are not the same. Was this change intended?

File: searchapp/utils/utils.py [hunk 12]
```diff
@@ -153,46 +164,55 @@ def convert_question_bank(question_bank_path):
...
             try:
-                subject = row[0].value.lower()
-                chapter_no = int(row[1].value)
-                chapter_name = row[2].value
-                question_type = row[3].value
-                question_type = str(int(question_type)) if (type(question_type) == type(1.0) or type(question_type) == type(1)) else question_type.lower()
-                question_text = row[4].value
-                question_source = row[5].value
-                subject_to_chapter_to_question[subject][(chapter_no, chapter_name)][question_type].append((question_text, question_source))
+                board = row[0].value
```
Comment (PaiAkshay998) [inline searchapp/utils/utils.py]: Use `lower` for text based stuff. Otherwise case will be an issue

Comment (PaiAkshay998) [reply searchapp/utils/utils.py]: Also, make sure that the updated question bank with board and grade are uploaded.

Comment (ishanisri) [reply searchapp/utils/utils.py]: Added this

File: searchapp/utils/utils.py (hunk 11 above)
Comment (ishanisri) [reply searchapp/utils/utils.py]: It was flaky as it did not work for me but worked for Sandhya. I have reverted this change now.

File: searchapp/utils/utils.py [hunk 13]
```diff
@@ -77,19 +77,23 @@ def convert_marker_data(marker_path, subject_breakup):
...
-                    questions_of_given_type.append((marks_secured, question_number, chapter_number))
+                    questions_of_given_type.append(
+                        (marks_secured, question_number, chapter_number))
                 questions_of_given_type.sort(key=lambda x: x[0], reverse=True)
                 questions_of_given_type = questions_of_given_type[:questions_to_attempt]
                 student_to_answer[student][question_type] = questions_of_given_type
                 question_type_start_row += total_questions
                 question_type = header_row[question_type_start_row].value
+                for val in chapter_row:
+                    print(type(val.value))
```
Comment (PaiAkshay998) [inline searchapp/utils/utils.py]: Remove this.

**Comment by ishanisri** [comment]
After checking in the master branch, the subject splits worked the same way as they are working now and not the way it is being pointed out to work. Hence , this PR had till now just focused on getting the old functionality working the same way as before after changes to the model.

**Comment by PaiAkshay998** [comment]
> After checking in the master branch, the subject splits worked the same way as they are working now and not the way it is being pointed out to work. Hence , this PR had till now just focused on getting the old functionality working the same way as before after changes to the model.

Seems to be working as pointed out itself.

2019-09-11T121352_1366x768
